packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -1641,20 +1641,30 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     _lastBottomViewInset = WidgetsBinding.instance.window.viewInsets.bottom;
   }
 
+  _WhitespaceDirectionalityFormatter _whitespaceFormatter;
+
   void _formatAndSetValue(TextEditingValue value) {
+    _whitespaceFormatter ??= _WhitespaceDirectionalityFormatter(textDirection: _textDirection);
+
     // Check if the new value is the same as the current local value, or is the same
     // as the post-formatting value of the previous pass.
     final bool textChanged = _value?.text != value?.text;
-    final bool isRepeat = value?.text == _lastFormattedUnmodifiedTextEditingValue?.text;
-    if (textChanged && !isRepeat && widget.inputFormatters != null && widget.inputFormatters.isNotEmpty) {
-      for (final TextInputFormatter formatter in widget.inputFormatters)
-        value = formatter.formatEditUpdate(_value, value);
-      _value = value;
-      _updateRemoteEditingValueIfNeeded();
-    } else {
+    final bool isRepeat = value == _lastFormattedUnmodifiedTextEditingValue;
+    // Always pass the text through the whitespace directionality formatter to maintain
+    // expected behavior with carets on trailing whitespace.
+    if (!isRepeat) {
+      if (textChanged && widget.inputFormatters != null && widget.inputFormatters.isNotEmpty) {
+        for (final TextInputFormatter formatter in widget.inputFormatters) {
```

Comment:
How did you decide whether user formatters should run before or after the correction?

Reply:
The reasoning for calling it after is that this would be less risky as the user's formatters always receive the raw untouched data. Any changes we make are strictly after at least one pass through any user formatters.

Reply:
sounds reasonable